File: java/doit_java/Chapter6/src/singleton/Company3.java
// public 메서드 만들기
package singleton;

public class Company3 {
    private static Company3 instance = new Company3();
    private Company3() {}

    // 인스턴스를 외부에서 참조할 수 있도록 public get() 메서드 구현
    public static Company3 getInstance() {
        if (instance == null)
            instance = new Company3();
        return instance;    // 유일하게 생성한 인스턴스 반환
    }
}

File: python/python_oop_practice/p_oop_2/chap05/staticex/company3.py
# 싱글톤(singleton) 구현 방법, Method 3: A base class
class Singleton(object):
  _instance = None
  def __new__(class_, *args, **kwargs):
    if not isinstance(class_._instance, class_):
        class_._instance = object.__new__(class_, *args, **kwargs)
    return class_._instance


class Company(Singleton):
  pass


def main():
    company1 = Company()
    company2 = Company()

    print(company1)
    print(company2)


if __name__ == '__main__':
    main()
